### PR TITLE
Feature/rolling update

### PR DESCRIPTION
**Pull request by lftakakura**
Deploy pods before deleting them in a rolling-update respecting maxSurge

### CODE REVIEW

**Comment by andrehp** [comment]
Would it add too much complexity to allow maxSurge and maxUnavailable on the same config like deployments do?

**Comment by cscatolini** [comment]
> Would it add too much complexity to allow maxSurge and maxUnavailable on the same config like deployments do?

I'd rather not change any behaviour and/or scheduler config at this moment @andrehp. Do you think this is a must have?

**Comment by andrehp** [comment]
> > Would it add too much complexity to allow maxSurge and maxUnavailable on the same config like deployments do?
> 
> I'd rather not change any behaviour and/or scheduler config at this moment @andrehp. Do you think this is a must have?

I wouldn't say it's a must have, but it might be nice to speed up deploys, though I do agree it might not be the best idea to change this behavior right now.

> I agree @cscatolini. Maestro already has the query string maxsurge which suffices many cases

I did a quick search on the docs about maxsurge but didn't find anything. But I think maestro's maxsurge is different from a k8s deployment's, isn't it?

File: controller/utils.go [hunk 1]
```diff
@@ -166,81 +178,65 @@ func rollback(
...
 	ticker := time.NewTicker(1 * time.Second)
 	defer ticker.Stop()
 
-	i := 0
 	for {
 		exit := true
 		select {
 		case <-ticker.C:
-			if operationManager.WasCanceled() {
+			if operationManager != nil && operationManager.WasCanceled() {
```
Comment (cscatolini) [inline controller/utils.go]: in which scenario operationManager would be nil?

File: controller/utils.go [hunk 2]
```diff
@@ -276,51 +271,83 @@ func waitTerminatingPods(
 		exit := true
 		select {
 		case <-ticker.C:
-			for _, pod := range deletedPods {
+			if operationManager.WasCanceled() {
```
Comment (cscatolini) [inline controller/utils.go]: shouldn't you have the same check here?
`if operationManager != nil && operationManager.WasCanceled() {`

File: controller/utils.go [hunk 3]
```diff
@@ -276,51 +271,83 @@ func waitTerminatingPods(
...
+						logger.
+							WithError(err).
+							WithField("pod", pod.GetName()).
+							Errorf("error recreating pod")
+					}
+					break
+				}
+
+				if len(createdPod.Status.Phase) == 0 {
+					//HACK! Trying to detect if we are running unit tests
```
Comment (cscatolini) [inline controller/utils.go]: can't this have any unintended side effects? Maybe add a logger here?

**Comment by cscatolini** [comment]
> > > Would it add too much complexity to allow maxSurge and maxUnavailable on the same config like deployments do?
> > 
> > 
> > I'd rather not change any behaviour and/or scheduler config at this moment @andrehp. Do you think this is a must have?
> 
> I wouldn't say it's a must have, but it might be nice to speed up deploys, though I do agree it might not be the best idea to change this behavior right now.

You can talk to @lftakakura. We don't want major changes but we believe that creating 25% new pods before deleting previous versions may be a big burden in our infrastructure. He is thinking about some way to speed the process and will update this PR soon.

> 
> > I agree @cscatolini. Maestro already has the query string maxsurge which suffices many cases
> 
> I did a quick search on the docs about maxsurge but didn't find anything. But I think maestro's maxsurge is different from a k8s deployment's, isn't it?

Yes. It is a global maestro's config instead of per scheduler. The default we're using in prod is 25%.

File: controller/utils.go (hunk 1 above)
Comment (lftakakura) [reply controller/utils.go]: operationManger is nil when rolling back (rollback can't be canceled)

File: controller/utils.go (hunk 2 above)
Comment (lftakakura) [reply controller/utils.go]: Yes it should

File: controller/utils.go [hunk 4]
```diff
@@ -166,81 +178,65 @@ func rollback(
...
-					logger.WithField("pod", pod.GetName()).Debugf("pod still exists")
-					exit = false
-					break
-				}
 
-				newPod, err := roomManager.Create(logger, mr, redisClient,
-					db, clientset, configYAML, scheduler)
-				if err != nil {
+				if err == nil || !strings.Contains(err.Error(), "not found") {
+					logger.WithField("pod", pod.GetName()).Debugf("pod still exists, deleting again")
```
[the remark this answers is not in the document]
Comment (lftakakura) [reply controller/utils.go]: Updated

File: controller/utils.go (hunk 2 above)
Comment (lftakakura) [reply controller/utils.go]: Updated

File: controller/utils.go [hunk 5]
```diff
@@ -39,41 +39,59 @@ func replacePodsAndWait(
...
 	scheduler *models.Scheduler,
 	operationManager *models.OperationManager,
 ) (createdPods []v1.Pod, deletedPods []v1.Pod, timedout, canceled bool) {
 	createdPods = []v1.Pod{}
 	deletedPods = []v1.Pod{}
 
-	for _, pod := range podsToDelete {
-		logger.Debugf("deleting pod %s", pod.GetName())
+	// create a chunk of pods (chunkSize = maxSurge)
+	for _, pod := range podsChunk {
```
Comment (ghostec) [inline controller/utils.go]: This way I think maxSurge is locked to 1 no matter what. What needs to be done is invoking the block inside this for in routines and waiting on all of them (using channel or sync.WaitGroup) to finish before continuing the loop

Comment (lftakakura) [reply controller/utils.go]: What you're suggesting is for each pod in the chunk to be created in parallel? If so, we could have the problem of creating too many pods at once

**Pull request merged**